rfc/composite_table_availability.pseudocode

```diff
@@ -0,0 +1,123 @@
+
+PhysicalTableSchema:
+	getFieldName(Column c)
+
+DataSource:
+	String getName()
+	Map<FieldName, Set<Segment>> raw segment data
+	Map<FieldName, List<Interval>> squashedFromSegments  // optional optimization
+
+	Set<FieldName> getFields() //union of all keys
+	List<Interval> getIntervals(FieldName fieldName)
+	Map<FieldName,List<Interval>> getIntervals(Set<FieldName>) // map slice from squashedMap
+	Set<SegmentId> getSegments(List<Interval>)  // Filter all segments by overlap on time
+
+DataSourceConstraints // Simple bean parameter object
+
+	getGrain
+	Set<FieldName> getMetrics
+	Set<Dimension> getDimensions
+	getApiFilters
+
+	DataSourceConstraints(ApiRequest)
+	DataSourceConstraints (DruidQuery, ApiFilter apifilters)
+
+	withMetrics(Set<FieldName> metrics) // Copy constructor
+
+
+DataSourceAvailabilityResolver:
+	List<String> getDataSourceNames(DataSourceConstraints)
```

Comment:
I'm still not sure I like returning strings instead of objects here (following the idea of "move to string land as late as possible"), but until we see this in action, I don't have any specific objections, since it may be that the only thing after this is called is serialization